### PR TITLE
Add Step, ELU, SELU, Swish activation functions

### PR DESCRIPTION
**Pull request by nsantavas**
This commit adds the Step, ELU, SELU and Swish activations functions

### CODE REVIEW

**Comment by awni** [comment]
Can you rebase on https://github.com/ml-explore/mlx/pull/108/files and remove the duplicate activations?

**Comment by nsantavas** [comment]
> Can you rebase on https://github.com/ml-explore/mlx/pull/108/files and remove the duplicate activations?

Done

**Comment by nsantavas** [comment]
@awni rebased again to main due to conflicts

**Comment by m0saan** [comment]
@nsantavas could you please add the new activation functions to the docs? https://github.com/ml-explore/mlx/blob/f4ddd7dc44aef7eb4d41b7f372d985e513ff808f/docs/src/python/nn.rst?plain=1#L161C2-L161C2

**Comment by nsantavas** [comment]
> @nsantavas could you please add the new activation functions to the docs? https://github.com/ml-explore/mlx/blob/f4ddd7dc44aef7eb4d41b7f372d985e513ff808f/docs/src/python/nn.rst?plain=1#L161C2-L161C2

@m0saan done

File: python/tests/test_nn.py [hunk 1]
```diff
@@ -378,6 +378,49 @@ def test_log_sigmoid(self):
...
+                -1.5201621055603027,
+                -1.1113275289535522,
+                0.0,
+                1.0506999492645264,
+                2.1013998985290527,
+                3.152099847793579,
+            ]
+        )
+        y = nn.SELU()(x)
+        self.assertTrue(mx.array_equal(y, expected))
```
Comment (awni) [inline python/tests/test_nn.py:405]: Use `mx.allclose` here as I've noticed exact equality sometimes fails due to subtle numerical differences.

File: python/tests/test_nn.py [hunk 2]
```diff
@@ -378,6 +378,49 @@ def test_log_sigmoid(self):
...
+                -0.26894140243530273,
+                0.0,
+                0.7310585975646973,
+                1.7615940570831299,
+                2.857722282409668,
+            ]
+        )
+        y = nn.Swish()(x)
+        mx.array_equal(y, expected)
+        self.assertTrue(mx.array_equal(y, expected))
```
Comment (awni) [inline python/tests/test_nn.py:422]: Same comment as above re `mx.allclose`. Also the first `array_equal` is not used, could you remove it?

File: python/mlx/nn/layers/activations.py [hunk 3]
```diff
@@ -143,6 +143,52 @@ class Sigmoid(Module):
...
+
+def swish(x):
+    r"""Applies the Swish Activation Function.
+
+    .. math::
+        \text{swish}(x) = x * \sigma(x)
+
+    where :math:`\sigma(\cdot)` is the logistic sigmoid.
+    """
+    return x * mx.sigmoid(x)
```
Comment (awni) [inline python/mlx/nn/layers/activations.py:189]: This is the same function as the `SiLU`. Could you remove it + the accompanying changes, and maybe just update the SiLU docstring to mention it is also known as the `swish`?

File: python/mlx/nn/layers/activations.py [hunk 4]
```diff
@@ -143,6 +143,52 @@ class Sigmoid(Module):
     pass
 
 
+def step(x: mx.array, threshold: float = 0.0):
+    r"""Applies the Step Activation Function.
```
Comment (awni) [inline python/mlx/nn/layers/activations.py:147]: nit: lower case "Activation Function."

File: python/mlx/nn/layers/activations.py [hunk 5]
```diff
@@ -274,3 +320,37 @@ def tanh(x):
 @_make_activation_module(tanh)
 class Tanh(Module):
     pass
+
+
+class Step(Module):
+    r"""Applies the Step Activation Function.
```
Comment (awni) [inline python/mlx/nn/layers/activations.py:326]: nit: lower case "Activation Function."

File: python/mlx/nn/layers/activations.py [hunk 6]
```diff
@@ -143,6 +143,52 @@ class Sigmoid(Module):
     pass
 
 
+def step(x: mx.array, threshold: float = 0.0):
```
Comment (awni) [inline python/mlx/nn/layers/activations.py:146]: This is a non-standard activation (given it's derivative is 0 almost everywhere). What's the motivation to include it?

**Review by awni — CHANGES_REQUESTED** [review]
Thanks for the contributions, this is looking good! A few more comments / questions.

**Comment by nsantavas** [comment]
@awni Done

**Review by awni — APPROVED** [review]

**Pull request merged**